Firestore/core/test/unit/bundle/bundle_serializer_test.cc

```diff
@@ -0,0 +1,114 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/bundle/bundle_serializer.h"
+
+#include "Firestore/Protos/cpp/firestore/bundle.pb.h"
+#include "google/protobuf/util/json_util.h"
+#include "gtest/gtest.h"
+
+namespace firebase {
+namespace firestore {
+namespace bundle {
+namespace {
+
+using google::protobuf::Message;
+using google::protobuf::util::MessageToJsonString;
+using nlohmann::json;
+using ProtoBundleMetadata = ::firestore::BundleMetadata;
+using util::ReadContext;
+
+ProtoBundleMetadata bundle_metadata() {
+  ProtoBundleMetadata proto_metadata{};
+  *proto_metadata.mutable_id() = "bundle-1";
+  proto_metadata.mutable_create_time()->set_seconds(2);
+  proto_metadata.mutable_create_time()->set_nanos(3);
+  proto_metadata.set_version(1);
+  proto_metadata.set_total_bytes(123456789987654321L);
+  proto_metadata.set_total_documents(9999);
+  return proto_metadata;
+}
+
+std::string replaced_copy(const std::string& source,
+                          const std::string& pattern,
+                          const std::string& value) {
+  std::string result{source};
+  auto start = result.find(pattern);
+  result.replace(start, pattern.size(), value);
+  return result;
+}
+
+TEST(BundleSerializerTest, DecodesBundleMetadata) {
+  auto proto_metadata = bundle_metadata();
+
+  std::string json_string;
+  MessageToJsonString(proto_metadata, &json_string);
+
+  BundleSerializer serializer;
+  ReadContext context;
+  BundleMetadata actual = serializer.DecodeBundleMetadata(context, json_string);
+
+  EXPECT_TRUE(context.ok());
+  EXPECT_EQ(proto_metadata.id(), actual.bundle_id());
+  EXPECT_EQ(proto_metadata.create_time().seconds(),
+            actual.create_time().timestamp().seconds());
+  EXPECT_EQ(proto_metadata.create_time().nanos(),
+            actual.create_time().timestamp().nanoseconds());
+  EXPECT_EQ(proto_metadata.version(), actual.version());
+  EXPECT_EQ(proto_metadata.total_bytes(), actual.total_bytes());
+  EXPECT_EQ(proto_metadata.total_documents(), actual.total_documents());
+}
+
+TEST(BundleSerializerTest, DecodesInvalidBundleMetadataReportsError) {
+  auto proto_metadata = bundle_metadata();
+
+  std::string json_string;
+  MessageToJsonString(proto_metadata, &json_string);
+  auto invalid = "123" + json_string;
+
+  BundleSerializer serializer;
+  ReadContext context;
+  serializer.DecodeBundleMetadata(context, invalid);
+
+  EXPECT_FALSE(context.ok());
+
+  // Replace total_bytes to a string unparseable to integer.
+  std::string json_copy =
+      replaced_copy(json_string, "123456789987654321", "xxxyyyzzz");
```

Comment:
Can you also test the case where one of the numbers is a string containing a number that is too large to fit into an integer (so that `SimpleAtoi` should return `false`)?

Reply:
Done.